src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure.VirtualChars;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.RoutePattern;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteHandlers;
+
+public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
+{
+    private static void DisallowNonParsableComplexTypesOnParameters(
+        in OperationAnalysisContext context,
+        WellKnownTypes wellKnownTypes,
+        IInvocationOperation invocation,
+        IMethodSymbol methodSymbol)
+    {
+        var routePatternArgument = invocation.Arguments[1];
+
+        if (routePatternArgument.Value.ConstantValue is not { HasValue: true } constant || constant.Value is not string routeTemplate)
+        {
+            return;
+        }
+
+        var enumerator = new RouteTokenEnumerator(routeTemplate);
+
+        while (enumerator.MoveNext())
+        {
+            var routeParameterName = enumerator.CurrentName.ToString();
+
+            var violatingParameter = methodSymbol.Parameters.FirstOrDefault(
+                p => p.Name == routeParameterName && !IsParsable(p.Type, wellKnownTypes)
+                );
```

Comment:
I believe the logic for IsParsable already exists in another analyzer. Check out:

https://github.com/dotnet/aspnetcore/blob/4ff55988a7f634aa443693bdd917488b9b018cd9/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs#L380-L455

I think this logic can be moved to a common location and shared between analyzers.

Reply:
Lifted this code into a Common namespace and transplanted a few of the types there. Take a look at let me know if that aligns with what you were thinking. I also extracted some of the methods that were already present around determining parsability and augmented them a little to handle the ```IParsable<TSelf>``` scenario.